src/spi/spi.hpp

```diff
@@ -0,0 +1,65 @@
+#ifndef SRC_SPI_SPI_HPP_
+#define SRC_SPI_SPI_HPP_
+
+#include <array>
+#include <spi_types.hpp>
+#include "spi_interface.hpp"
+#include "stm32g4xx_hal.h"
+
+//preserve include order
+#include "spi_config.h"
+
+namespace spi {
+
+///Maximum buffer size allowed for SPI Transaction
+static constexpr std::uint8_t SPI_TRANSACTION_BUFFER_SIZE_LIMIT = 64;
+static constexpr std::uint8_t SPI_HAL_TX_RX_TIMEOUT = 100;
+
+//Valid values for chip select pin setting
+enum class PinSetting : bool {
+  HIGH = true,
+  LOW = false
+};
+
+//Data structure to hold the chip select GPIO information
+typedef struct CSPinDefinition {
+  GPIO_TypeDef *peripheral;
+  uint16_t gpio_pin;
+} CSPin;
+
+/**
+ * @brief Concrete implementation of the SPI interface.
+ * 
+ */
+class SPI final : spi::SPIInterface {
+ public:
+  SPI() = delete;
+  explicit SPI(const CSPin chip_select) : spi::SPIInterface(), chip_select_(chip_select){};
+  virtual ~SPI() = default;
+
+  auto Transfer(std::vector<uint8_t> &miso_data_buffer, std::vector<uint8_t> &mosi_data_buffer) noexcept -> spi::SPIStatus override;
```

Comment:
Could you describe the Interface of Transfer here a little bit?
I'm not so familiar with SPI. So as a user of this interface, which buffer have I to use to send and receive data?
How big should they be. You use resize() inside of the method. So is the size of miso_data_buffer somehow relevant?

Reply:
Transfer() is a pure virtual member of the spi_interface. I will add some more description there. Will be a separate issue.